### PR TITLE
Add axes argument to ndimage filters

### LINKED ISSUE #10554: ndimage.gaussian_filter provide axis option
**Is your feature request related to a problem? Please describe.**
I have a stacked array of 2d images, which is a 3d array. I want to use ndimage.gaussian_filter for each 2d image in the array. However, current method does 3d gaussian, which I did not realize at the beginning. 

**Describe the solution you'd like**
Will the Scipy provides an option for axis and dimension in ndimage.gaussian_filter? It would be much more convenient and faster than doing it a loop. Thanks!

### PR DESCRIPTION
**Pull request by grlee77**
closes #10554
see also requests/discussion at: https://github.com/scipy/scipy/issues/12997#issuecomment-811268933 and https://github.com/scipy/scipy/issues/16191#issuecomment-1144781807

cc @mdhaber 

# description

This PR adds an "axes" keyword argument to most `scipy.ndimage` filters (1D filters already have an `axis` argument so this PR applies only to the others). Having an `axes` keyword-only argument allows easier filtering of a subset of the axes of an image. For example to apply a 2D footprint along 2 axes of an n-dimensional array, the user could either:
1.) expand the footprint to size one along the other axes. 
or 
2.) applied for loop(s) over non-filtered axes
However, this is a bit error prone and not always obvious to users, so I think it is helpful to provide this via an `axes` argument.

The functions for which a new `axes` argument can be introduced are:
- convolve
- correlate
- gaussian_filter
- gaussian_gradient_magnitude
- gaussian_laplace
- generic_gradient_magnitude
- generic_laplace
- laplace
- maximum_filter
- median_filter
- minimum_filter
- percentile_filter
- rank_filter
- uniform_filter
- generic_filter

For now, I reverted back to including only `gaussian_filter` in this PR. I can push other filters in a follow up.

This could potentially be extended to morphology functions (e.g. `erode`, `dilate`).

### CODE REVIEW

**Comment by mdhaber** [comment]
@grlee77 If I'm to review, please let's start with one small PR on one function. It will make review easier and speed things up overall to do one PR per function (at least initially). Once we get the hang of it, perhaps we can do multiple functions in a PR, but let's start by working incrementally. Thanks!

File: scipy/ndimage/_filters.py
Comment (mdhaber) [inline scipy/ndimage/_filters.py:314]: Presumably the `i`th element of these arguments corresponds with the `i`th axis in `axes`? I think that is worth mentioning explicitly, and perhaps a test can checks this (if it's not already)?

File: scipy/ndimage/_ni_support.py
Comment (mdhaber) [inline scipy/ndimage/_ni_support.py:117]: Will this miss when the same axis is referred to be a negative and positive index?

Comment (mdhaber) [reply scipy/ndimage/_ni_support.py:117]: No. I missed `axes = tuple(ax % ndim if ax < 0 else ax for ax in axes)` before.

File: scipy/ndimage/_ni_support.py
Comment (mdhaber) [inline scipy/ndimage/_ni_support.py:106]: I would have suggested `Sequence` since order matters, but apparently an `ndarray` isn't a `Sequence`, so I suppose this is fine.

File: scipy/ndimage/_ni_support.py
Comment (mdhaber) [inline scipy/ndimage/_ni_support.py:104]: There was a pretty long discussion about the best way to do this in gh-7351. It looked like there was a slight preference for using `operator.index`, but I leave that to your discretion.

File: scipy/ndimage/tests/test_filters.py
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py:782]: Was it intentional to leave out a test with `order` here?

**Review by mdhaber — COMMENTED** [review]
This looks really good. Let's just trim the PR to `gaussian_filter` and send a message to the mailing list. Let's give the opportunity for feedback until next weekend, and after that I think we can merge this. I think we can merge follow-up PRs quite quickly; there's no need to wait for feedback on each unless there is other feedback about this one.

File: scipy/ndimage/_filters.py
Comment (mdhaber) [inline scipy/ndimage/_filters.py:1052]:
```suggestion
```
I assume this was for debugging?

```
scipy.ndimage.uniform_filter
----------------------------

File "build/testenv/lib/python3.9/site-packages/scipy/ndimage/_filters.py", line 1085, in uniform_filter
Failed example:
    result = ndimage.uniform_filter(ascent, size=20)
Expected nothing
Got:
    axes=[(0, 20, 0, 'reflect'), (1, 20, 0, 'reflect')]
```

File: scipy/ndimage/_filters.py
Comment (grlee77) [reply scipy/ndimage/_filters.py:314]: Yes. Initially I was returning `sorted(axes)` from `_check_axes`, but then I would have to also sort any of these other tuples in the same way which seemed more complicated. I can update the description.

**Comment by grlee77** [comment]
> This looks really good. Let's just trim the PR to gaussian_filter and send a message to the mailing list. Let's give the opportunity for feedback until next weekend, and after that I think we can merge this. 

Okay, I will trim it back to just `gaussian_filter` for now and send a message to the mailing list tomorrow.

File: scipy/ndimage/_ni_support.py
Comment (grlee77) [reply scipy/ndimage/_ni_support.py:106]: I would also be okay with being strict about only supporting `tuple`, but this way is more in line with how the rest of the arguments are currently handled in the `ndimage` module. (`_normalize_sequence` in this same file also uses `Iterable`)

Comment (mdhaber) [reply scipy/ndimage/_ni_support.py:106]: I agree. Maybe in SciPy 2.0 we can make this sort of thing strict, but I think for now we probably should be flexible.

File: scipy/ndimage/_ni_support.py
Comment (grlee77) [reply scipy/ndimage/_ni_support.py:104]: I will check it using `operator.index`

File: scipy/ndimage/tests/test_filters.py
Comment (grlee77) [reply scipy/ndimage/tests/test_filters.py:782]: no, I can add it

File: scipy/ndimage/_filters.py
Comment (grlee77) [reply scipy/ndimage/_filters.py:1052]: thanks. planning to revert the non-Gaussian changes for now, so the suggestion will become outdated

File: scipy/ndimage/tests/test_filters.py
Comment (grlee77) [reply scipy/ndimage/tests/test_filters.py:719]: order is now tested here as well

File: scipy/ndimage/_filters.py
Comment (mdhaber) [inline scipy/ndimage/_filters.py:311]: Nit: the axes themselves will not be filtered.
```suggestion
        If None, `input` is filtered along all axes. Otherwise,
```

File: scipy/ndimage/_filters.py
Comment (mdhaber) [inline scipy/ndimage/_filters.py:315]:
```suggestion
        corresponds to the ith entry in `axes`.
```

File: scipy/ndimage/_filters.py
Comment (mdhaber) [inline scipy/ndimage/_filters.py:312]:
```suggestion
        `input` is filtered along the specified axes. When `axes` is
```

File: scipy/ndimage/_ni_support.py
Comment (mdhaber) [inline scipy/ndimage/_ni_support.py:115]: Consider the Oxford comma to avoid confusion between "(iterable of integers) or (None)" and "iterable of (integers or None)". 

You're welcome to maintain the formatting if you prefer, but while I'm at it, I might as well suggest:
```suggestion
        message = "axes must be an integer, iterable of integers, or None"
        raise ValueError(message)
```

**Review by mdhaber — APPROVED** [review]
I've played with this locally and it works as advertised. Everything looks correct and the tests are strong. I would probably parameterize the last test over `kwargs` for concision, `match` the error messages in the tests, and in `test_gauss_axes_kwargs` the assembly of `XYZ_3d` could be simpler with arrays/array indexing, but these are just make-better. I haven't reviewed for `ndimage` before, so I'll respect the customs of the subpackage : )

Something optional to consider while we're here is the Note about `radius`. It's `radius or (sigma*truncate)`, really.

The refguide/doc CI failure is real, but once it's happy, I approve. I'll wait until next weekend to allow time for participation from the mailing list. After that, we can make our way through the other functions quickly. Feel free to combine functions if the tests can be combined.

Note for future PRs - if it's OK with you, let's avoid force-push and I'll just squash-merge unless there's really important history to preserve. By just adding commits on top, it's easier to see what has changed between commits.

**Comment by grlee77** [comment]
Thanks for reviewing @mdhaber! I just sent out to the scipy-dev list for feedback today, so let's give it another week or so. I will make the suggested revisions here in the meantime.

File: scipy/ndimage/tests/test_filters.py
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py:783]: Just a note that this is what I had in mind with the parameterization.
```python3
    @pytest.mark.parametrize('tuple_ndim', [1, 3])
    @pytest.mark.parametrize('mismatch_val',
                             [('radius', 3), ('mode', 'constant'),
                              ('sigma', 1.0), ('order', 0)])
    def test_gauss_axes_tuple_length_mismatch(self, tuple_ndim, mismatch_val):
        array = numpy.arange(6 * 8 * 12, dtype=numpy.float64).reshape(6, 8, 12)
        kwargs = dict(sigma=1.0, radius=(3, 3), axes=(0, 1), mode='constant')
        mismatched, value = mismatch_val
        kwargs[mismatched] = (value,) * tuple_ndim
        err_msg = "sequence argument must have length equal to input rank"
        with pytest.raises(RuntimeError, match=err_msg):
            ndimage.gaussian_filter(array, **kwargs)
```

**Comment by mdhaber** [comment]
Thanks @grlee77! I can take a look at the next function when you're ready! If there's one that can reuse some of the same tests w/ `@parametrize`, that would be great (not sure).

**Pull request merged**
